actors/builtin/miner/miner_state.go

```diff
@@ -1014,123 +1022,83 @@ func (st *State) PenalizeFundsInPriorityOrder(store adt.Store, currEpoch abi.Cha
 // Unlocks an amount of funds that have *not yet vested*, if possible.
 // The soonest-vesting entries are unlocked first.
 // Returns the amount actually unlocked.
-func (st *State) UnlockUnvestedFunds(store adt.Store, currEpoch abi.ChainEpoch, target abi.TokenAmount) (abi.TokenAmount, error) {
-	vestingFunds, err := adt.AsArray(store, st.VestingFunds)
-	if err != nil {
-		return big.Zero(), err
-	}
-
+func (st *State) UnlockUnvestedFunds(currEpoch abi.ChainEpoch, target abi.TokenAmount) (abi.TokenAmount, error) {
 	amountUnlocked := abi.NewTokenAmount(0)
-	lockedEntry := abi.NewTokenAmount(0)
-	var toDelete []uint64
-	var finished = fmt.Errorf("finished")
-
-	// Iterate vestingFunds are in order of release.
-	err = vestingFunds.ForEach(&lockedEntry, func(k int64) error {
-		if amountUnlocked.LessThan(target) {
-			if k >= int64(currEpoch) {
-				unlockAmount := big.Min(big.Sub(target, amountUnlocked), lockedEntry)
-				amountUnlocked = big.Add(amountUnlocked, unlockAmount)
-				lockedEntry = big.Sub(lockedEntry, unlockAmount)
-
-				if lockedEntry.IsZero() {
-					toDelete = append(toDelete, uint64(k))
-				} else {
-					if err = vestingFunds.Set(uint64(k), &lockedEntry); err != nil {
-						return err
-					}
-				}
-			}
-			return nil
-		} else {
-			return finished
+	lastIndex := -1
+
+	for i := range st.VestingFunds {
+		if amountUnlocked.GreaterThanEqual(target) {
+			break
 		}
-	})
 
-	if err != nil && err != finished {
-		return big.Zero(), err
+		vf := st.VestingFunds[i]
+		if vf.Epoch >= currEpoch {
+			vestedAmt := vf.Amount
+
+			unlockAmount := big.Min(big.Sub(target, amountUnlocked), vestedAmt)
+			amountUnlocked = big.Add(amountUnlocked, unlockAmount)
+			newAmount := big.Sub(vestedAmt, unlockAmount)
+
+			if newAmount.IsZero() {
+				lastIndex = i
+			} else {
+				st.VestingFunds[i].Amount = newAmount
+			}
+		}
 	}
 
-	err = vestingFunds.BatchDelete(toDelete)
-	if err != nil {
-		return big.Zero(), errors.Wrapf(err, "failed to delete locked fund during slash: %v", err)
+	// remove all entries upto and including lastIndex
+	if lastIndex != -1 {
+		st.VestingFunds = st.VestingFunds[lastIndex+1:]
```

Comment:
This will remove the vested funds we just skipped.

Reply:
@Stebalien I'm not sure what you mean. `lastIndex` will point to the last fund we zeroed. We want to remove all entries upto and including that .

If funds for an epoch are zero here, it follows that all funds before it will be zero as well. What am I missing here ?

Reply:
We may have funds that _should_ have vested but haven't been popped from the queue yet. We'll zero out all epochs skipped by https://github.com/filecoin-project/specs-actors/pull/843/files#diff-2580e4b82238370ee854a52e8d5428e7R1055.

Reply:
@anorth Is this still an issue ?

Reply:
If that's the case, we need to change this method to assert that there are no funds in the queue that should have already been unlocked (instead of just skipping them).